### PR TITLE
BCF-2685: move types from core to support Solana plugin command extraction

### PR DESCRIPTION
**Pull request by jmank88**
https://smartcontract-it.atlassian.net/browse/BCF-2685

Supports:
- https://github.com/smartcontractkit/chainlink-solana/pull/550
- https://github.com/smartcontractkit/chainlink/pull/10973

### CODE REVIEW

File: pkg/loop/config.go
Comment (jmank88) [inline pkg/loop/config.go:9]: My instinct was to expose this as mutable so that `package main`s can make their own decisions about what it should be, but maybe this is not necessary :thinking:

File: pkg/loop/prom.go
Comment (jmank88) [inline pkg/loop/prom.go:35]: I wanted to avoid exporting this, since the common helper code can handle it internally, but there is still one test in core using it. We may be able to do the same validation indirectly though :thinking:

File: pkg/loop/config.go
Comment (krehermann) [reply pkg/loop/config.go:9]: the plugin can't choose the value - it must be guaranteed to be unique and only the core node has enough context to do that

File: pkg/loop/prom.go
Comment (krehermann) [reply pkg/loop/prom.go:35]: The core test is exercising the loop registration and service discovery.

I think we need that test, because is it is only test for the metric plumbing between the two sides.

Part of the contract btw the core node and a LOOPP is that a LOOPP ought to expose a prom server. We could promote that to a real interface somehow and then mock on the core side test for the registry.

or this repo could expose a mock/test implementation that runs a prom server that could be used for that test. 

or we could leave this public and keep the test as is.

i think 1 is the most principled thing to do, but doesn't fit in this PR, so maybe 2 is the best immediate option?

File: pkg/loop/prom_test.go
Comment (krehermann) [inline pkg/loop/prom_test.go:47]: why is this defined in a test file? is that an oversight?

File: pkg/loop/config.go
Comment (jmank88) [reply pkg/loop/config.go:9]: >  it must be guaranteed to be unique and only the core node has enough context to do that.

That's why I think it's important to be mutable. And I'm trying to avoid making assumptions about the plugin implementations from here. A practical case could be if we want to write a test host for some reason.

File: pkg/loop/prom_test.go
Comment (jmank88) [reply pkg/loop/prom_test.go:47]: It was only used in tests unless I missed something.

File: pkg/loop/config.go
Comment (jmank88) [reply pkg/loop/config.go:9]: Oh but we can choose to not inherit the host env with newer hashicorp, so maybe this is nbd.

Comment (jmank88) [reply pkg/loop/config.go:9]: The name of the env var. If there was a collision to avoid, or a different standard to follow than `CL_` prefixing everything, then this would need to be mutable for that to be possible. But we don't want to inherit the host env when launching a plugin anyways, so collisions don't really matter in practice, and the `CL_` standard doesn't even need to apply here. Maybe we should actually considering using something else, e.g. `LOOPP_` :shrug:

File: pkg/loop/prom_test.go
Comment (jmank88) [reply pkg/loop/prom_test.go:47]: I'll connect the other layers in drafts before merging this to be sure :shrug:

**Review by patrickhuie19 — DISMISSED** [review]

File: pkg/loop/config.go
Comment (jmank88) [inline pkg/loop/config.go:17]: Something felt off about the old API, and I think it's because this type was only an interface because it was capturing existing config types previous (for log fields), but that is no longer the case. A struct keeps things much simpler and more extensible (we can add fields without breaking users).

Comment (jmank88) [reply pkg/loop/config.go:17]: Also allowed for simplifying the helpers to methods: `(*EnvConfig).SetCmdEnv`

File: pkg/loop/prom.go
Comment (jmank88) [inline pkg/loop/prom.go:35]: @krehermann WDYT about dropping the functional options? I couldn't find satisfactory names (old ones weren't qualified enough in the context of this broader package), and there were extra allocations with the default handler. I'm open to other ideas :shrug:

File: pkg/loop/prom.go
Comment (jmank88) [reply pkg/loop/prom.go]: Exported again.

File: pkg/loop/adapter.go
Comment (jmank88) [reply pkg/loop/adapter.go]: Indeed, this was unused :shrug:

File: pkg/loop/plugin.go
Comment (jmank88) [inline pkg/loop/plugin.go:12]: This was called `plugins.Base` before.

File: pkg/loop/prom.go
Comment (jmank88) [reply pkg/loop/prom.go:35]: Oh the allocations is a separate problem.

Still not sure about the naming for functional opts though. `WithHandler` is too general for this package namespace, and even `WithPromHandler` has the same problem. So then is `WithPromServerHandler` the best option? That would mean that callers have to do:
`loop.NewPromServer(port, lggr, loop.WithPromServerHandler(handler))`
With an extra `loop.WithPromServer` per option, compared to a struct+method.

File: pkg/loop/config.go
Comment (jmank88) [inline pkg/loop/config.go:13]: WDYT? @krehermann @patrickhuie19

Comment (patrickhuie19) [reply pkg/loop/config.go:13]: What's the problem we're solving? Distinguishing between the environment of the host and plugins?

**Review by patrickhuie19 — APPROVED** [review]

File: pkg/loop/config.go
Comment (jmank88) [reply pkg/loop/config.go:13]: Namespacing so we don't have collisions. We do this in core. These aren't user facing so it's not as important, but in theory we could have a conflict with a dependency that has it's own env vars :shrug:

Comment (patrickhuie19) [reply pkg/loop/config.go]: Hmm - If we prefix everything with `CL_`, then we're still at risk of collisions.

**Review by patrickhuie19 — APPROVED** [review]

File: pkg/loop/config.go
Comment (jmank88) [reply pkg/loop/config.go]: Indeed, but they are less likely :shrug:

**Review by patrickhuie19 — APPROVED** [review]

File: pkg/loop/prom.go
Comment (krehermann) [reply pkg/loop/prom.go:35]: sorry, seems like being tagged isn't reach me clear thru github. i didn't realize you had asked my opinion here. aside: how do you handle that? it seems to be lost in a stream of other review comment in an email chain for me.

i find options creating an instance confusing. it's easier for me to understand for the constructor of the thing to take options.

is the only use case for testing right now? 

what about something like 

// WithHandler uses the given HTTP handler. It must be called before Start to have any effect
func (*PromServer) WithHandler(h) *PromServer

and the usage would be ~
s := NewPromServer(port,logger).WithHandler(h)

Comment (jmank88) [reply pkg/loop/prom.go:35]: > sorry, seems like being tagged isn't reach me clear thru github. i didn't realize you had asked my opinion here. aside: how do you handle that? it seems to be lost in a stream of other review comment in an email chain for me.

Mentions come through my email stream :shrug: 

> i find options creating an instance confusing. it's easier for me to understand for the constructor of the thing to take options.

Can you elaborate on this? I find the opposite, because functional options come with a ton of extra baggage and behave like 'spooky action at a distance'. They strike me as particularly odd when they are only setting fields rather than doing something complex or dynamic.

A `(*PromServer).WithHandler` method seems cleaner than options to me, but it still doesn't seem like it's in the right place if it comes with a caveat that you can only call it at certain times. A basic func or method constructor doesn't come with any extra baggage or caveats - it just does exactly what we need and reads very plainly.

Do you find `net.ListenConfig` difficult to read? https://pkg.go.dev/net#ListenConfig

**Comment by jmank88** [comment]
Merged to unblock the growing stack of PRs, but I'm happy to circle back on anything.

**Pull request merged**
